### ISSUE TITLE
Adding the Edge features in GIN (GINE)

### ISSUE THREAD

**Issue by thegodone** [akensert/molgraph#5]
Can you add the Edge features in GIN to get the so called GINE version based on this paper https://arxiv.org/pdf/1905.12265.pdf ?
equation 2.1 page 3 and more exactly equation A.1 (without last Relu) page 16
A nice implementation is available from there: https://github.com/yuyangw/MolCLR/blob/3d3bc1912be27b0c97435fd9134f4d4c73d4c5ab/models/ginet_molclr.py

**Comment by akensert** [comment]
There are too many GNN variants out there.. :D 

That said, I will try to implement a GIN-E variant (based on the paper you reference) in the near future. One difference is that I think I will keep the `node_feature_update = (1 + eps) * prev_node_feature + node_feature_aggregated` part of the GIN model which seems to be omitted in the GIN-E implementation you reference? Not sure if this is the right choice, thoughts? 

After that I will look into AttentiveFP. I've seen it mentioned and studied in several papers now, so I will try to implement that one too in the near future.

As for DMPNN/DGIN conv layers that you requested previously: I found it difficult to design it the way we want it to be --- namely, in a way so that we can just replace e.g. `MPNNConv` with `DMPNNConv` in a keras sequential model). The main reason is that these models start with an **"initial projection"** (matrix transform) before the message passing steps, and ends with a **"final projection"** after the last message passing step. And for DGIN, **node aggregation steps are performed subsequently**. I guess additional arguments could be added to these layers (e.g. initialize_edge_state=True to the first layer, and perform_node_readout=True on the last layer), but I wasn't satisfied with this implementation/design. Instead, I'm planning to implemented a new layer: `EdgeConv` which can be used in this way:

(Note that `EdgeConv` is significantly different from the other gnn layers as it passes message based on edges instead of nodes. Also note that the `node_feature` and `edge_feature` field of the outputted graph tensor are unchanging, while an `edge_state` field is initialized and updated instead.)

DMPNN:
```
dmpnn_model = tf.keras.Sequential([
  EdgeConv(units=128),
  EdgeConv(units=128),
  NodeReadout(target='edge_state'),
  Readout()
])
```

DGIN:
```
dgin_model = tf.keras.Sequential([
  EdgeConv(units=128),
  EdgeConv(units=128),
  NodeReadout(target='edge_state'),
  GINConv(),                         # node aggregation steps
  GINConv(),                         # node aggregation steps
  Readout(),
])
```

This implementation will differ in some ways from the DMPNN and DGIN model implementations. The models uses the initial node states as residuals in the aggregation/propagation steps, while for the implementation above will use the very previous node states as residuals instead (both for the `EdgeConv` part as well as the node aggregaton part (`GINConv`). However, I think what is most important here is the `EdgeConv`. 

Let me know what you think.

Finally, I'm also adding a feature to the `MPNNConv` (as well as the `EdgeConv`) layer to allow weight tying (see `MPNNConv` documentation later when it's implemented), which is used in both in the original MPNN implementaton as well as DMPNN and DGIN (as far as I remember/understand). In other words, the different aggregation steps of these models shares the same kernel/weight matrix.

**Comment by thegodone** [comment]
For GIN/E: it was shown that generaly epsilon is set to zero! but you can do clearly node_feature_update = (1 + eps) * prev_node_feature + node_feature_aggregated => and eps is set by default to zero.
EdgeConv is a good idea: are you able to do directed ? 

FInally, in almost all recent graphs you have an initial embedding layer for both edge and node features to be sure we have the same hidden dimension for the rest of the graph operation like concate, add etc...

sharing the same matrix at each convolution may be an option yes, but this is less important. I like the edge_state idea. 

You will see that in AttentiveFP there are two type of convolution other the nodes and other the molecule (super node). This is also the same for several recent models.

Can you write a true example to use ESOL data on the current code you provide model.DMPNN or DGIN I have issue to work with it.

**Comment by akensert** [comment]
About DMPNN/DGIN/EdgeConv: I'm trying to figure out how to deal with subgraphs without edges (right now it raises an error). Should I enforce self loops? Or perhaps do a conditional where only subgraphs with edges passes through the `message_passing`/`updates` and the subgraphs without edges just skips it and gets assigned a zero tensor?

I've also found a bug unrelated to DMPNN/DGIN/EdgeConv. I'm trying to fix that first. In brief: in relatively rare cases, where `node_feature` and `edge_feature` have the same number of rows (namely, there are equal number of nodes and edges), graph_tensor.separate() separates `edge_feature` incorrectly; and `tf.concat([graph_tensor_1, graph_tensor_2, ...], axis=0)` gets messed up too (as it has a similar conditional statement in it where it checks if a tensor matches nodes or edges based on its size/number of rows). 

Let's see how much I can manage (in addition to fixing bugs). This is quite time consuming unfortunately :( `EdgeConv` is almost ready so I should be able to push that one soon.

**Comment by akensert** [comment]
@thegodone I've now added `EdgeConv`/`NodeReadout` implementations. I've also fixed some important bugs. Please update molgraph to the latest version.
